src/throttled-queue.js

```diff
@@ -0,0 +1,120 @@
+ */
+function getOrigin(url) {
+    if (!url) {
+        return '';
+    }
+    const origin = (new URL(url)).origin;
+    if (origin.endsWith('.whatwg.org')) {
+        return 'whatwg.org';
+    }
+    else if (origin.endsWith('.github.io')) {
+        return 'github.io';
+    }
+    else if (origin.endsWith('.csswg.org') ||
+             origin.endsWith('.css-houdini.org') ||
+             origin.endsWith('.fxtf.org')) {
+        return 'csswg.org';
+    }
+    else {
+        return origin;
+    }
+}
+
+
+/**
+ * The ThrottledQueue class can be used to run a series of tasks that send
+ * network requests to an origin server in parallel, up to a certain limit,
+ * while guaranteeing that only one request will be sent to a given origin
+ * server at a time.
+ */
+module.exports = class ThrottledQueue {
+  originQueue = {};
+  maxParallel = 4;
+  ongoing = 0;
+  pending = [];
+
+  constructor(maxParallel) {
+    if (maxParallel >= 0) {
+      this.maxPar = maxParallel;
+    }
+  }
+
+  /**
+   * Run the given processing function with the given parameters, immediately
+   * if possible or as soon as possible when too many tasks are already running
+   * in parallel.
+   *
+   * Note this function has no notion of origin. Users may call the function
+   * directly if they don't need any throttling per origin.
+   */
+  async runThrottled(processFunction, ...params) {
+    if (this.ongoing >= this.maxParallel) {
+      return new Promise((resolve, reject) => {
+        this.pending.push({ params, resolve, reject });
+      });
+    }
+    else {
+      this.ongoing += 1;
+      const result = await processFunction.call(null, ...params);
+      this.ongoing -= 1;
+
+      // Done with current task, trigger next pending task in the background
+      setTimeout(_ => {
+        if (this.pending.length && this.ongoing < this.maxParallel) {
+          const next = this.pending.shift();
+          this.runThrottled(processFunction, ...next.params)
+            .then(result => next.resolve(result))
+            .catch(err => next.reject(err));
+        }
+      }, 0);
+
+      return result;
+    }
+  }
+
+  /**
+   * Run the given processing function with the given parameters, immediately
+   * if possible or as soon as possible when too many tasks are already running
+   * in parallel, or when there's already a task being run against the same
+   * origin as that of the provided URL.
+   *
+   * Said differently, the function serializes tasks per origin, and calls
+   * "runThrottled" to restrict the number of tasks that run in parallel to the
+   * requested maximum.
+   *
+   * Additionally, the function forces a 2 second sleep after processing to
+   * keep a low network profile.
+   */
+  async runThrottledPerOrigin(url, processFunction, ...params) {
+    const origin = getOrigin(url);
+    if (!this.originQueue[origin]) {
+      this.originQueue[origin] = Promise.resolve(true);
+    }
+    return new Promise((resolve, reject) => {
```

Comment:
Using `Promise.withResolvers` would probably help with readability

Reply:
Never heard of that ;)

But this ([baseline 2024](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers)) function was apparently added in Node.js 22.0.0, not in Node.js LTS for now!